File: src/main/java/com/example/fastpass/model/Recarga.java
package com.example.fastpass.model;

import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.OneToOne;
import java.time.LocalDate;

/**
 * Representa o evento de recarregar o passe. Toda recarga é feita através
 * de um Pagamento (Pix ou Debito) — a associação é 1 para 1, conforme o
 * diagrama de classes.
 */
@Entity
public class Recarga {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private double valor;

    private LocalDate data;

    @Enumerated(EnumType.STRING)
    private StatusRecarga status;

    @OneToOne
    @JoinColumn(name = "pagamento_id")
    private Pagamento pagamento;

    protected Recarga() {
        // construtor padrão exigido pelo JPA
    }

    public Recarga(double valor, LocalDate data, Pagamento pagamento) {
        this.valor = valor;
        this.data = data;
        this.pagamento = pagamento;
        this.status = StatusRecarga.PENDENTE;
    }

    /**
     * Confirma a recarga processando o pagamento vinculado.
     * RN03: uma recarga só é confirmada após a validação do pagamento simulado.
     */
    public boolean confirmar() {
        if (pagamento != null && pagamento.processarPagamento()) {
            this.status = StatusRecarga.CONFIRMADA;
            return true;
        }
        this.status = StatusRecarga.CANCELADA;
        return false;
    }

    public Long getId() {
        return id;
    }

    public double getValor() {
        return valor;
    }

    public LocalDate getData() {
        return data;
    }

    public StatusRecarga getStatus() {
        return status;
    }

    public Pagamento getPagamento() {
        return pagamento;
    }
}

File: frontend/fast_pass_poo/lib/models/recarga.dart
enum FormaPagamento { PIX, DEBITO }

enum StatusRecarga { PENDENTE, CONFIRMADA, CANCELADA }

class Recarga {
  final int id;
  final double valor;
  final String data;
  final StatusRecarga status;
  final double saldoAtual;

  Recarga({
    required this.id,
    required this.valor,
    required this.data,
    required this.status,
    required this.saldoAtual,
  });

  factory Recarga.fromJson(Map<String, dynamic> json) {
    return Recarga(
      id: json['id'],
      valor: (json['valor'] as num).toDouble(),
      data: json['data'],
      status: StatusRecarga.values.byName(json['status']),
      saldoAtual: json['saldoAtual'] != null
          ? (json['saldoAtual'] as num).toDouble()
          : 0.0,
    );
  }
}